New Spreedly gateway for PaywayDotCom

DanAtPayway opened the pull request:
Hello,

I added this pull for the New Spreedly gateway for PayWayDotCom.

Let me know if you need our PaywayWS Document.

The reponse.message will contain our 4 digit paywayCode concatenated with a '-' (dash) character followed by a paywayMessage (text description that is could change in the future).  The 4 digit paywayCode will not change.  Both are described in our PaywayWS Credit Card Integration Guide. 

Dan

tatsianaclifton commented:
@DanAtPayway I noticed that implementation is for supporting other payment types besides credit cards. However, all tests are written just for credit cards. Could you please add unit and remote tests for all supported payment types or change implementation to support just credit cards? I will continue with code review once this is addressed. Thank you

DanAtPayway commented:
Tatsiana,

I checked in the updated code,

Dan

tatsianaclifton commented on lib/active_merchant/billing/gateways/payway_dot_com.rb:
```diff
@@ -0,0 +1,268 @@
+        response['paywayCode'] == PAYWAY_WS_SUCCESS
+      end
+
+      def error_code_from(response)
+        if success_from(response)
+          ''
+        else
+          error = !STANDARD_ERROR_CODE_MAPPING[response['paywayCode']].nil? ?
+          STANDARD_ERROR_CODE_MAPPING[response['paywayCode']] :
+          STANDARD_ERROR_CODE[:processing_error]
+        end
+      end
```
Is there a reason to set error_code to empty string versus just keep it nil? Anyway I think the method can be simplified to something like 
```suggestion
      def error_code_from(response)
        return '' if success_from(response)
        
        error = !STANDARD_ERROR_CODE_MAPPING[response['paywayCode']].nil? ? 
        STANDARD_ERROR_CODE_MAPPING[response['paywayCode']] : 
        STANDARD_ERROR_CODE[:processing_error]
        return error
      end
```

Or can just do `return if success_from(response)` if there is no specific reason to set error to empty string

tatsianaclifton commented on lib/active_merchant/billing/gateways/payway_dot_com.rb:
```diff
@@ -0,0 +1,268 @@
+        end
+      end
+
+      def message_from(success, response)
+        if !response['paywayCode'].nil?
+          if success
+            return response['paywayCode'] + "-" + "success"
+          else
+            return response['paywayCode'] + "-" + response['paywayMessage']
+          end
+        end
+        ""
+      end
```
Maybe it can be simplified to something like 
```suggestion
      def message_from(success, response)
        return '' if response['paywayCode'].nil?

        return response['paywayCode'] + '-' + 'success' if success

        response['paywayCode'] + '-' + response['paywayMessage']
      end
```
Or can just do `return if response['paywayCode'].nil?` if there is no specific reason to set message to an empty string.

tatsianaclifton commented on lib/active_merchant/billing/gateways/payway_dot_com.rb:
```diff
@@ -0,0 +1,268 @@
+        ""
+      end
+
+      def authorization_from(response)
+        if success_from(response)
+          if !response['cardTransaction'].nil?
+            return response['cardTransaction']['name'] if response['cardTransaction']['name']
+          end
+        end
+        ""
+      end
```
Maybe something simpler will also work
```suggestion
      def authorization_from(response)
        return '' unless success_from(response) && response['cardTransaction'].nil?

        response['cardTransaction']['name']
      end
```

Can be just `return unless success_from(response) && response['cardTransaction'].nil?` if no specific reason to set authorization to an empty string.

tatsianaclifton reviewed (changes requested):
Hello,

Thanks for working on that implementation. Overall, it looks very good so far. Great job! Just couple questions and small suggestions. Once fixed I'll look over one more time to make sure we don't have any more questions before merging.

DanAtPayway commented:
Tatsiana,

I have completed all the changes you requested.

I changed idSource to sourceId, removed unused code and comments and updated code per suggestions.

The Authorization Hold PaywayWS request is "authorization", not "authorizationSale" as you saw in the document/guide.  We have updated our document to reflect this.

Dan

tatsianaclifton reviewed (approved):
Looks good overall. Approved, but I will make some styling changes and merge after that.
